org.eclipse.scanning.sequencer/src/org/eclipse/scanning/sequencer/nexus/NexusScanFileManager.java

```diff
@@ -244,62 +244,69 @@ public NexusScanInfo getNexusScanInfo() {
 		
 		// add the global metadata scannables, and the required metadata scannables for
-		// each scannable in the scan
-		metadataScannableNames.addAll(deviceConnectorService.getGlobalMetadataScannableNames());
+		// each scannable in the scan. These are from the legacy GDA8 location map
+		perScanMonitorNames.addAll(deviceConnectorService.getGlobalMetadataScannableNames());
 		
 		// the set of scannable names to check for dependencies
 		Set<String> scannableNamesToCheck = new HashSet<>();
-		scannableNamesToCheck.addAll(metadataScannableNames);
+		scannableNamesToCheck.addAll(perScanMonitorNames);
 		scannableNamesToCheck.addAll(scannableNames);
 		do {
 			// check the given set of scannable names for dependencies
 			// each iteration checks the scannable names added in the previous one
-			Set<String> requiredScannables = scannableNamesToCheck.stream().flatMap(
-					name -> deviceConnectorService.getRequiredMetadataScannableNames(name).stream())
-					.filter(name -> !metadataScannableNames.contains(name))
+			Set<String> requiredScannables = scannableNamesToCheck.stream()
+					.flatMap(name -> deviceConnectorService.getRequiredMetadataScannableNames(name).stream())
+					.filter(name -> !perScanMonitorNames.contains(name))
 					.collect(Collectors.toSet());
 			
-			metadataScannableNames.addAll(requiredScannables);
+			perScanMonitorNames.addAll(requiredScannables);
 			scannableNamesToCheck = requiredScannables;
 		} while (!scannableNamesToCheck.isEmpty());
 		
-		// remove any scannable names in the scan from the list of metadata scannables
-		metadataScannableNames.removeAll(scannableNames);
-		
-		// get the metadata scannables for the given names
-		final List<IScannable<?>> monitors = model.getMonitors() != null ? model.getMonitors() : new ArrayList<>();
-		for (String scannableName : metadataScannableNames) {
-			IScannable<?> metadataScannable = deviceConnectorService.getScannable(scannableName);
-			try {
-			    metadataScannable.setMonitorRole(MonitorRole.PER_SCAN);
-			} catch (IllegalArgumentException ne) {
-				continue;
+		// remove any scannable names in the scan from the list of per scan monitors,
+		// as a scannable can only have one role within the scan
+		perScanMonitorNames.removeAll(scannableNames);
+		
+		// now that we've calculated the set of names for the per point monitors in the scan,
+		// remove all the existing monitors, so we've just got the ones we want to add
+		perScanMonitorNames.removeAll(existingPerScanMonitorNames);
+		if (!perScanMonitorNames.isEmpty()) {
+			// get the monitors with these names and add them to the list, provided they can be set to be per scan
+			final List<IScannable<?>> monitors = new ArrayList<>(model.getMonitors());
+			for (String scannableName : perScanMonitorNames) {
+				if (!existingPerScanMonitorNames.contains(scannableName)) {
+					IScannable<?> metadataScannable = deviceConnectorService.getScannable(scannableName);
+					try {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS134)